syntax = "proto3";

package pujpb;

enum EntryCategory {
  EC_NONE = 0;
  EC_COLLOQUIAL = 1;
  EC_LITERARY = 2;
  EC_CONVENTIONAL = 3;
}

enum EntryFrequency {
  EF_COMMON = 0;
  EF_LESS_COMMON = 1;
  EF_RARE = 2;
  EF_VERY_RARE = 3;
}

message EntryDetailExample {
  string teochew = 1;
  optional string puj = 2;
  optional string mandarin = 3;
}

message EntryDetail {
  // 释义
  optional string meaning = 1;
  // 例词
  repeated EntryDetailExample examples = 2;
}

message Pronunciation {
  // 声母
  string initial = 1;
  // 韵母
  string final = 2;
  // 音调
  int32 tone = 3;
}

message Entry {
  uint32 index = 1;
  // 繁体字
  string char = 2;
  // 简体字
  string char_sim = 3;
  // 发音
  Pronunciation pron = 4;
  // 字音类型
  EntryCategory cat = 5;
  // 字音使用频率
  EntryFrequency freq = 6;
  // 参考字（讹用字、错读字、训读字、异体字，等等）
  optional string char_ref = 7;
  // 字音详细信息
  repeated EntryDetail details = 8;
  // 可读鼻化的语音点
  repeated string accents_nasalized = 9;

  // 各口音点又音
  message PronunciationAka {
    string accent_id = 1;
    repeated Pronunciation prons = 2;
    // 是否替换原读音，默认为否，即又音，但又音与原读音并存，不挤兑原读音
    bool replace = 3;
  }
  repeated PronunciationAka pron_aka = 10;
}

message Entries {
  repeated Entry entries = 1;
}
